lib/dot.js

```diff
@@ -113,4 +98,176 @@ export default class Dot extends TileLayer.Canvas {
+    const p23 = this.getGridValue(x2, y3)
+    
+    const p30 = this.getGridValue(x3, y0)
+    const p31 = this.getGridValue(x3, y1)
+    const p32 = this.getGridValue(x3, y2)
+    const p33 = this.getGridValue(x3, y3)
+    
+    return new BicubicPatch (p00, p01, p02, p03, p10, p11, p12, p13, p20, p21, p22, p23, p30, p31, p32, p33)
+  }
+      
+}
+
+// An instance of this class represents a curve fit to four uniformly spaced data points.
+// Given four adjacent values a, b, c, d, it provides interpolated values between b and c using a and d to 
+// determine the slope going into and out of the b-c interval.
+class CubicHermiteInterpolator {
+    
+  constructor (a, b, c, d) {
+    this.c3 = -a / 2.0 + (3.0 * b) / 2.0 - (3.0 * c) / 2.0 + d / 2.0
+    this.c2 = a - (5.0 * b) / 2.0 + 2.0 * c - d / 2.0
+    this.c1 = -a / 2.0 + c / 2.0;
+    this.c0 = b;
+  }
+
+  // Given a value in [0, 1] expressing the position between b and c, return the interpolated value.
+  interpolate (fraction) {
+    return this.c3 * fraction ** 3 + this.c2 * fraction ** 2 + this.c1 * fraction + this.c0;
+  }
+     
+}
+
+// Represents a pre-calculated interpolation for a patch of sixteen data samples.
+// For a 4x4 grid of samples, this allows us to calculate interpolated values between the central four samples.
+// By pre-fitting the curves in one dimension (y) and proceeding with the interpolation row by row, 
+// we re-use most of the computation from one output pixel to the next.
+class BicubicPatch {
+
+  // Supply an unrolled row-major grid of 16 values (a 4x4 grid).
+  // The resulting object can be used to interpolate between the inner four cells.
+  // Maybe we should be initializing this with a typed 2D array instead of this mess of individual variables.
+  constructor (p00, p01, p02, p03, p10, p11, p12, p13, p20, p21, p22, p23, p30, p31, p32, p33) {
+    // Create interpolations through each of the four columns
+    this.columnInterpolator0 = new CubicHermiteInterpolator(p00, p01, p02, p03)
+    this.columnInterpolator1 = new CubicHermiteInterpolator(p10, p11, p12, p13)
+    this.columnInterpolator2 = new CubicHermiteInterpolator(p20, p21, p22, p23)
+    this.columnInterpolator3 = new CubicHermiteInterpolator(p30, p31, p32, p33)
+  }
+
+  // Perform curve fitting in the second (x) dimension based on the pre-fit curves in the y dimension.
+  setFractionY (yFraction) {
+    const p0 = this.columnInterpolator0.interpolate(yFraction)
+    const p1 = this.columnInterpolator1.interpolate(yFraction)
+    const p2 = this.columnInterpolator2.interpolate(yFraction)
+    const p3 = this.columnInterpolator3.interpolate(yFraction)
+    this.rowInterpolator = new CubicHermiteInterpolator(p0, p1, p2, p3) 
+  }
+
+  getInterpolatedValue (xFraction) {
+    return this.rowInterpolator.interpolate(xFraction)
+  }
+  
+}
+
+class InterpolatingColorScale {
+
+  // Pass in an array of [[z, r, g, b, a], [z, r, g, b, a], [z, r, g, b, a]...]
```

Comment:
Yes, that's correct. `d3-color` parses many color formats and normalizes them. It's not so much a compatibility thing as a consistency with other code thing, for instance https://github.com/conveyal/analysis-ui/blob/dev/lib/components/analysis/spectrogram.js#L24